src/Validation/Rule/NoUnusedVariablesRule.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Digia\GraphQL\Validation\Rule;
+
+use Digia\GraphQL\Error\ValidationException;
+use Digia\GraphQL\Language\AST\Node\NodeInterface;
+use Digia\GraphQL\Language\AST\Node\OperationDefinitionNode;
+use Digia\GraphQL\Language\AST\Node\VariableDefinitionNode;
+use Digia\GraphQL\Language\AST\Node\VariableNode;
+use function Digia\GraphQL\Validation\unusedVariableMessage;
+
+/**
+ * No unused variables
+ *
+ * A GraphQL operation is only valid if all variables defined by an operation
+ * are used, either directly or within a spread fragment.
+ */
+class NoUnusedVariablesRule extends AbstractRule
+{
+    /**
+     * @var array|VariableDefinitionNode[]
+     */
+    protected $variableDefinitions;
+
+    /**
+     * @inheritdoc
+     */
+    public function enterNode(NodeInterface $node): ?NodeInterface
+    {
+        if ($node instanceof OperationDefinitionNode) {
+            $this->variableDefinitions = [];
+        }
+
+        if ($node instanceof VariableDefinitionNode) {
+            $this->variableDefinitions[] = $node;
+        }
+
+        return $node;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function leaveNode(NodeInterface $node): ?NodeInterface
+    {
+        if ($node instanceof OperationDefinitionNode) {
+            $variableNamesUsed = [];
+            $usages            = $this->validationContext->getRecursiveVariableUsages($node);
+            $operationNameNode = $node->getName();
+            $operationName     = null !== $operationNameNode ? $operationNameNode->getValue() : null;
+
+            /** @var VariableNode $variableNode */
+            foreach ($usages as ['node' => $variableNode]) {
```

Comment:
This syntax looks weird, are you sure this is correct?

Reply:
This is correct, it's the destructor syntax.